src/processor/fingerprint.cpp

```diff
@@ -0,0 +1,283 @@
+    value_hash_field &operator=(value_hash_field &&) = default;
+
+    std::size_t length() override { return value.type == DDWAF_OBJ_MAP ? 8 : 0; }
+    void operator()(string_buffer &output) override;
+
+    ddwaf_object value;
+};
+
+template <typename T, typename... Rest>
+std::size_t generate_fragment_length(T &generator, Rest... rest)
+{
+    if constexpr (sizeof...(rest) > 0) {
+        return generator.length() + 1 + generate_fragment_length(rest...);
+    } else {
+        return generator.length();
+    }
+}
+
+template <typename T, typename... Rest>
+void generate_fragment_field(string_buffer &buffer, T &generator, Rest... rest)
+{
+    generator(buffer);
+    if constexpr (sizeof...(rest) > 0) {
+        buffer.append('-');
+        generate_fragment_field(buffer, rest...);
+    }
+}
+
+template <typename... Args>
+ddwaf_object generate_fragment(std::string_view header, Args... generators)
+{
+    std::size_t total_length = header.size() + 1 + generate_fragment_length(generators...);
+
+    string_buffer buffer{total_length};
+    buffer.append_lowercase(header);
+    buffer.append('-');
+
+    generate_fragment_field(buffer, generators...);
+
+    ddwaf_object res;
+    auto [ptr, size] = buffer.move();
+    ddwaf_object_stringl_nc(&res, ptr, size);
+
+    return res;
+}
+
+// Retur true if the first argument is less than (i.e. is ordered before) the second
+bool str_casei_cmp(std::string_view left, std::string_view right)
+{
+    auto n = std::min(left.size(), right.size());
+    for (std::size_t i = 0; i < n; ++i) {
+        auto lc = ddwaf::tolower(left[i]);
+        auto rc = ddwaf::tolower(right[i]);
+        if (lc != rc) {
+            return lc < rc;
+        }
+    }
+    return left.size() <= right.size();
+}
+
+void normalize_string(std::string_view key, std::string &buffer, bool trailing_separator)
+{
+    // Clear should not deallocate...
+    // TODO: verify
+    buffer.clear();
+
+    if (buffer.capacity() < key.size()) {
+        // Add space for the extra comma, just in case
+        buffer.reserve(key.size() + 1);
+    }
+
+    for (auto c : key) {
+        if (c == ',') {
+            buffer.append(R"(\,)");
+        } else {
+            buffer.append(1, ddwaf::tolower(c));
+        }
+    }
+
+    if (trailing_separator) {
+        buffer.append(1, ',');
+    }
+}
+
+void string_hash_field::operator()(string_buffer &output)
+{
+    if (value.empty()) {
+        return;
+    }
+
+    cow_string value_lc{value};
+    transformer::lowercase::transform(value_lc);
+
+    sha256_hash hasher;
+    hasher << static_cast<std::string_view>(value_lc);
+
+    hasher.write_digest(output.subspan<8>());
+}
+
+void key_hash_field::operator()(string_buffer &output)
+{
+    if (value.type != DDWAF_OBJ_MAP or value.nbEntries == 0) {
+        return;
+    }
+
+    std::vector<std::string_view> keys;
+    keys.reserve(value.nbEntries);
+
+    for (unsigned i = 0; i < value.nbEntries; ++i) {
```

Comment:
We're only doing a top-level scanning? What happens if the body has a map under a top level parameter?

Reply:
We did define this as only "top-level" fields, I'm not sure we should have a special case for a single map as we can't know for sure if that's always the schema of the body or just a one off.